src/dpapi_ng/_client.py

```diff
@@ -509,6 +572,94 @@ def ncrypt_unprotect_secret(
     return _decrypt_blob(blob, rk)
 
 
+def ncrypt_protect_secret(
+    data: bytes,
+    sid: string,
+    server: t.Optional[str] = None,
+    username: t.Optional[str] = None,
+    password: t.Optional[str] = None,
+    auth_protocol: str = "negotiate",
+    cache: t.Optional[KeyCache] = None,
+) -> bytes:
+    """Encrypt DPAPI-NG Blob.
+
+    Encrypts the blob provided as DPAPI-NG Blob. This is meant to
+    replicate the Win32 API `NCryptProtectSecret`_.
+
+    Encrypting the DPAPI-NG blob requires making an RPC call to the domain
+    controller for the domain the blob was created in. It will attempt this
+    by looking up the DC through an SRV lookup but ``server`` can be specified
+    to avoid this SRV lookup.
+
+    The RPC call requires the caller to authenticate before the key information
+    is provided. This user must be one who is authorized to encrypt the secret.
+    Explicit credentials can be specified, if none are the current Kerberos
+    ticket retrieved by ``kinit`` will be used instead. Make sure to install
+    the Kerberos extras package ``dpapi-ng[kerberos]`` to ensure Kerberos auth
+    can be used.
+
+    Args:
+        data: The bytes blob to encrypt.
+        server: The domain controller to lookup the root key info.
+        username: The username to encrypt the DPAPI-NG blob as.
+        password: The password for the user.
+        auth_protocol: The authentication protocol to use, defaults to
+            ``negotiate`` but can be ``kerberos`` or ``ntlm``.
+        cache: Optional cache that is used as the key source to avoid making
+            the RPC call.
+
+    Returns:
+        bytes: The encrypted DPAPI-NG data.
+
+    Raises:
+        ValueError: An invalid data structure was found.
+        NotImplementedError: An unknown value was found and has not been
+            implemented yet.
+
+    _NCryptProtectSecret:
+        https://learn.microsoft.com/en-us/windows/win32/api/ncryptprotect/nf-ncryptprotect-ncryptprotectsecret
+    """
+    root_key_identifier = None
+    key_identifier_l0 = -1
+    key_identifier_l1 = -1
+    key_identifier_l2 = -1
+
+    sd = sd_to_bytes(
+        owner="S-1-5-18",
+        group="S-1-5-18",
+        dacl=[ace_to_bytes(sid, 3), ace_to_bytes("S-1-1-0", 2)],
+    )
+
+    cache = cache or KeyCache()
+    rk = cache._get_key(
```

Comment:
I don't think we would ever hit a cache entry here as the root_key_identifier is `None`. It does open up an interesting question of how we can expose such information to the caller, should they be able to provide a custom rkid to suppose key generation from a cache. For now we can probably just remove the whole caching implementation and add that in a future version.

Reply:
[My suggestion](https://github.com/schorschii/dpapi-ng/commit/2693d8bda31949e1ea3581b7f378e6ec4c57cb5c) is to return a cache entry even if `root_key_id` is `None` and `l0`, `l1`, `l2` is `-1`. Only the security descriptor must match.

Reply:
It just doesn't make sense to have a cache if the root key id is not known. This id is critical to generating the proper seed values as it's used as part of the KDF context value. We won't know that root key id until the `GetKey` call is made which is after the cache is needed or the user explicitly provides a root key id.

Reply:
OK, for now I added a kwarg so that the user can specify the root_key_id.